File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int top = 0;
        int bottom = matrix.size() - 1;

        while (top <= bottom) {
            int midInd = floor((top + bottom) / 2);
            vector midMat = matrix[midInd];

            if (midMat.front() > target) {
                bottom = midInd - 1;
                continue;
            }
            if (midMat.back() < target) {
                top = midInd + 1;
                continue;
            }

            int l = 0;
            int r = midMat.size() - 1;

            while (l <= r) {
                int mid = floor((l + r) / 2);

                if (midMat[mid] == target) return true;
                else if (midMat[mid] < target) l = mid + 1;
                else r = mid - 1;
            }
            break;
        }
        return false;
    }
};

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        rows, cols = len(matrix), len(matrix[0])
        top, bot = 0, rows - 1

        while top <= bot:
            mid = (top + bot) // 2
            row = matrix[mid]

            if target > row[-1]: top = mid + 1
            elif target < row[0]: bot = mid - 1
            else: break
        
        if not (top <= bot):
            return False
        
        row = matrix[(top + bot) // 2]
        l = 0
        r = cols - 1
        while l <= r:
            mid = (l + r) // 2
            midVal = row[mid]

            if midVal == target: return True
            elif midVal < target: l = mid + 1
            else: r = mid - 1
        return False